### ISSUE TITLE
Question about using the I-ADOPT framework for matching aggregating parameters to variables in data files

### ISSUE THREAD

**Issue by gwemon** [i-adopt/questions#1]
As part of our implementation on the NVS of the I-ADOPT framework, we are working on the use case which aims to match aggregating parameters to instances of variable names held in multiple parameter labelling schemes and used to label data in various known data standards. For this, we are using the I-ADOPT ontology properties e.g. iop:hasProperty, iop:hasObjectOfInterest,  iop:hasMatrix to match the aggregation parameter to its relevant variables.
However, in many cases, this requires us to link one aggregated parameter to multiple applicable "properties" or multiple "Object of Interests" (i.e. all the ones that could qualify). 
See for example: 
http://vocab.nerc.ac.uk/collection/A05/current/EV_CO2/?_profile=iop&_mediatype=text/html
or
http://vocab.nerc.ac.uk/collection/A05/current/EV_NUTS/?_profile=iop&_mediatype=text/html
Is this an acceptable use of the IOP ontological properties?
If not, what would be our options?

**Comment by gwemon** [comment]
One suggestion I have would be to extend the ontology with the properties "hasApplicableProperties" "hasApplicableMatrices" etc...

**Comment by SirkoS** [comment]
Maybe to start, let me rephrase the issue to make sure I understand it in the correct way:

> For some variable the description is so generic that Property abd/or OoI can not be mapped to a single atomic component but you need a range of them (the nutrient example). A similar case exists when there is no generic concept available equivalent to the description but a range of more specialized ones are (the carbonate system example).

Both cases remind me of what we discussed as a hierarchy over variable descriptions and components. What seems to be missing is a more general concept to summarize some of the current atomic concepts. E.g., for the nutrient example, one could imagine a hierarchy for the nutrients like this
```
Nutrient            --- missing concept right now
┕╸ S27:CS026903     silicate
┕╸ S27:CS002877     nitrite
┕╸ S27:CS026904     phosphate
┕╸ S27:CS026905     nitrate+nitrite
┕╸ S27:CS002879     nitrate 
```

Using something like this, the variable can be rewritten as
```
A05:EV_NUTS
    hasMatrix             S21:S21S027	  water body
    hasObjectOfInterest   <someId>      Nutrient
    hasProperty           S06:S0600045  Concentration
```
and we're well in line with the framework as is. The carbon system example can be worked out in the same way.

If there is a need, one can also define more specialized variables
```
<someId>   "Concentration of Organic Nitrate"
    skos:broader          A05:EV_NUTS   Nutrients                      <--- hierarchy here
    hasMatrix             S21:S21S027	  water body
    hasObjectOfInterest   S27:CS002879  nitrate                        <--- specialized concept here
    hasProperty           S06:S0600045  Concentration
```

To me that would be the clean solution, sticking to the formal definition of the ontology while still detailing the information as much as possible for a given variable.


------------------

Sidenote:

I'm also somewhat uneasy with the mixing of different properties/quantities: In the carbonate system example, right now these properties are given:
* S06:S0600045	Concentration
* S06:S0600281	pH (unspecified scale)
* S06:S0600280	pH (NBS scale)
* S06:S0600047	Partial pressure
* S06:S0600277	pH (total scale)
* S06:S0600278	pH (free scale)
* S06:S0600279	pH (seawater scale)

One may aggregate all variations of pH into a single concept as before, but concentration and pressure are different quantities. I would not try to mix the remaining three properties (pH, concentration, and pressure) into a single concept. Such a concept has to real world equivalent for me. I assume you derived those properties from datasets annotated with this variable. To me that's actually a sign that the variable may not be defined in a proper way, leading to misuse by the scientists. However, the description does not mention concentration and pressure at all, so I'm not sure about the rationale here. From a scientists perspective those measurements might be "close enough", but for reuse and especially automated processing, such mix-ups will cause serious harm.

**Comment by huberrob** [comment]
Dear Gwen, 

I think you are semantically overloading the given quite generic term with meanings that are not identifiable from the term itself and I am not sure if one should use I-ADOPT for this.

For example adding 'hasMatrix' 'water body' to 'Carbonate system' is only possible because you know that the term is used in AtlantOS context which is a project related to marine research. In this context it is a 'seawater CO2–carbonate system' but without knowing this very specific context 'Carbonate system' can have a lot of meanings.

Further, 'Carbonate system' is used in AtlantOS context as 'essential variable' which in most cases represent a generic and flexible container for other variables. Therefore an essential variable likely this is different from the concept of an iop:Variable. 

I would recommend not to directly use I-ADOPT for 'essential variable' concepts, instead you should relate e.g. Seadatanet P01 variables and use I-ADOPT to describe each P01 variable individually.

Another comment: you added two different quantity kinds to the 'Carbonate system' but can a distinct variable have two different units? Is e.g. something measured in g/l or g/kg the same thing?

Robert

**Comment by gwemon** [comment]
Dear @SirkoS and @huberrob  Many thanks for your comments. I will try to respond to them in separate comments below. I think we do agree that the way we have currently used I-ADOPT to connect aggregating concepts to variables is not satisfactory. The key question is: how to connect "variables" (i.e. "I-ADOPT variables" like e.g. a P01 concept or a CF standard name) to an aggregating concept like the ones held in the A05 vocabulary other than by direct mapping? Can we use the decomposition promoted by the I-ADOPT framework in order to do so? If so what do we need to add to the framework to enable us to do so?

**Comment by gwemon** [comment]
My response/comments to @SirkoS comments: 
> Both cases remind me of what we discussed as a hierarchy over variable descriptions and components. What seems to be missing is a more general concept to summarize some of the current atomic concepts. E.g., for the nutrient example, one could imagine a hierarchy for the nutrients like this
> 
> ```
> Nutrient            --- missing concept right now
> ┕╸ S27:CS026903     silicate
> ┕╸ S27:CS002877     nitrite
> ┕╸ S27:CS026904     phosphate
> ┕╸ S27:CS026905     nitrate+nitrite
> ┕╸ S27:CS002879     nitrate 
> ```
> 
> Using something like this, the variable can be rewritten as
> 
> ```
> A05:EV_NUTS
>     hasMatrix             S21:S21S027	  water body
>     hasObjectOfInterest   <someId>      Nutrient
>     hasProperty           S06:S0600045  Concentration
> ```
> 
> and we're well in line with the framework as is. The carbon system example can be worked out in the same way.
> 
> If there is a need, one can also define more specialized variables
> 
> ```
> <someId>   "Concentration of Organic Nitrate"
>     skos:broader          A05:EV_NUTS   Nutrients                      <--- hierarchy here
>     hasMatrix             S21:S21S027	  water body
>     hasObjectOfInterest   S27:CS002879  nitrate                        <--- specialized concept here
>     hasProperty           S06:S0600045  Concentration
> ```
> 
> To me that would be the clean solution, sticking to the formal definition of the ontology while still detailing the information as much as possible for a given variable.

The problem here is that the concept "nutrient" means different things for different communities and contexts. While we can decide/agree/know what the essential variable EV_NUTS in A05 need to map to in terms of chemical entities (e.g. nitrate, nitrite, phosphate, silicate), the concept "nutrient" taken in isolation could be mapped to many more (or different) sets of chemical entities in a different context. And this is generally the case with aggregating terms. Because they are simpler, less detailed concepts, they do tend to be more dependent on context.

**Comment by gwemon** [comment]
Also, my response to @huberrob comments below:
> For example adding 'hasMatrix' 'water body' to 'Carbonate system' is only possible because you know that the term is used in AtlantOS context which is a project related to marine research. In this context it is a 'seawater CO2–carbonate system' but without knowing this very specific context 'Carbonate system' can have a lot of meanings.

That's correct. Concepts held in [A05](https://vocab.nerc.ac.uk/collection/A05/) are context specific. A05 is defined as a: Collection of terms used to group key measurements into a set of Essential Variables (EV) and their associated units as part of the AtlantOS Atlantic Ocean dataset 

> Further, 'Carbonate system' is used in AtlantOS context as 'essential variable' which in most cases represent a generic and flexible container for other variables. Therefore an essential variable likely this is different from the concept of an iop:Variable.

I agree. An essential variable is an aggregator (an aggregation concept) of I-ADOPT-type variables (like e.g. CF standard names or P01 concepts)

> I would recommend not to directly use I-ADOPT for 'essential variable' concepts, instead you should relate e.g. Seadatanet P01 variables and use I-ADOPT to describe each P01 variable individually.

Yes we are doing this indeed. We are expressing P01 and CF standard names using I-ADOPT too but in order to enable on-the-fly aggregation we also need to find a way of connecting aggregation concepts to their relevant I-ADOPT atoms.  

> Another comment: you added two different quantity kinds to the 'Carbonate system' but can a distinct variable have two different units? Is e.g. something measured in g/l or g/kg the same thing?

No it isn't the same quantity kind but in operational oceanography quantities dissolved in seawater can be expressed per unit volume or per unit mass of seawater depending on the community. And people often need to convert from one to another using (often) an assumed or a fixed density hence why we have to take this into consideration.

**Comment by gwemon** [comment]
So I totally agree that EVs are not valid I-ADOPT variables and hence we should not use the existing I-ADOPT properties to connect EV concepts to the I-ADOPT classes. EVs are aggregating concepts. However aggregation is one of the I-ADOPT use cases so we need to find a solution. Do we need to extend the current ontology to accommodate aggregation?

**Comment by mabablue** [comment]
interesting discussion here and sorry I am so late replying (too much to do as always). I do agree that we don't have a solution for this problem and it is worth thinking how to aggregate these variables per EV or how we should go forward.  think we should have an I-ADOPT meeeting to clarify this together. There is also another issue I will post we need to find a good solution for. So I will send a doodle for a 2 hours meeting to work on the two issues,

**Comment by kitchenprinzessin3880** [comment]
AtlantOS Essential Variables contains a mix of variables (can be 'atomised' based on the I-ADOPT definition) and 'groups' of variables. So mapping the latter (groups) to iop:variable will be overdesigned in my opinion. e.g., radiative flux this does not provide any context which can be further elaborated in terms of i-adopt concepts. maybe we can leave these 'groups' as just skos: concepts and then link them to i-adopt variables either through skos:broader (as @SirkoS  suggested) or new relation e.g. hasVariableKind (similar to qudt:hasQuantityKind)?

**Comment by gwemon** [comment]
Taken forward as: https://github.com/i-adopt/ontology/issues/8

**Comment by dr-shorthair** [comment]
Sorting and grouping (e.g. Nutrients) is likely to be application specific. 
I don't expect a single scheme to be useful across all applications.

**Comment by gwemon** [comment]
@dr-shorthair that's correct and this is why we are adopting a flexible approach to aggregation, enabling users to aggregate ("package") base variable concepts like e.g. P01 concepts or CF Standard Names in a way that fits a given application or given user group.

**Comment by graybeal** [comment]
I hope you don't mind if I reformulate at length some accumulated thoughts. If the answer to the next paragraph is 'no', then the rest should be ignored.

It seems to me the problem statement is "We need to be able to clearly define relationships of a collection of user- or context-specific entities to an aggregate entity". The naive case being a simple grouping like 'myNutrientsList'; the rich case being formulating the definition of a complex measurement or observation from its component parts. Do I understand the requirement correctly?

This seems to me to be an ontological task; in fact it seems to be the same as ontological modeling, in the same way the original I-ADOPT work was ontological modeling, but only to a specific level of complexity. (The idea that a CF term is naturally simpler than an A05 term is foreign to me. Some of the CF terms are wickedly rich in their implicit relations.)  Is the issue perhaps that you are dealing with the more detailed cases that didn't get addressed by round 1 of I-ADOPT modeling?

To take an example, when I look at the definition of http://vocab.nerc.ac.uk/collection/A05/current/EV_CO2, I see a whole host of implications and assumptions and special knowledge that are not at all pinned down by the definition itself. If you want a 'parameterization' of this concept, your choices are (A) a composition of all the terms that meet each existing relationship criteria (this is your starting point, and IMHO of course you can aggregate multiple objects satisfying the same property relation; but you can mostly use this to satisfy searches, not to understand or reason over your constructs);
(B) the inclusion of hierarchies and a refinement of some number of those relationships, so that you distinguish between (or narrow) the roles: total inorganic carbon and carbon dioxide must play different roles in the EV_CO2 model we are describing, and could be connected using different properties, potentially resulting in hierarchical or other ordered relations of the whole (option B is not necessarily a thorough solution and involves deciding about a great number of properties involved in scientific models); or
(C) a complete ontological specification of the relationships and objects throughout the model allowing you to describe any aggregating concept. This is an ontology itself, so maybe you haven't obtained much generalization by creating the framework around it, but it is a tremendous organization of knowledge.

When I was trying to crack this nut with CF concepts, the farthest I could imagine getting was that for the concrete list of standard names in CF, I could imagine building all the nested relationships that were necessary to define those standard names using concept collections from different vocabularies, that could be assembled erector-set style. But to describe all the terms that might be in any given scientific context (including the ever-more-aggregated ones that you're running into), I thought I'd have to assemble all the nouns, adverb and adjective and verb property relations, and probably other things that expressed the fundamental underlying relationship models. If only I had another month… 

So I think you are several levels deeper than I was in product maturity. But my question is, are you hoping to just improve it some (in which case an option B may be achievable and useful), or do you want to produce all the relations and noun sets that are needed for the erector set that will theoretically enable C, but could in the real world at least create a foundation for constructing such things on the fly?  In either case I think you are extending the number and composability of your relations; it is a matter of degree of difficulty needed to satisfy your most important use cases.

**Comment by gwemon** [comment]
Thank you @graybeal for your comments and questions. In your 3rd paragraph quoted below:

"This seems to me to be an ontological task; in fact it seems to be the same as ontological modeling, in the same way the original I-ADOPT work was ontological modeling, but only to a specific level of complexity. (The idea that a CF term is naturally simpler than an A05 term is foreign to me. Some of the CF terms are wickedly rich in their implicit relations.) Is the issue perhaps that you are dealing with the more detailed cases that didn't get addressed by round 1 of I-ADOPT modeling?"

I would have thought that the reason for our extension here is the opposite. i.e. we have dealt with the most detailed in the first phase and we are now dealing with the less detailed aspect of the way users refer to "variables". CF Standard Names, P01 PUV terms, EnvThes concepts for example are the most detailed terms, the ones that sit directly next to the result (number, text string, whatever it is) in data files and data bases. However, in order to build applications that gives access to data files that use these detailed concepts we need to find a way to link them to concepts that have been mainly developed for discovery and data products. These latter concepts are by design coarser and could potentially cover many of the more detailed concepts.

**Comment by graybeal** [comment]
Yes, if by 'detailed' you mean 'narrow' I agree with you, you have focused on detailed work.

The composition can take place in one of two ways: with precision, or without. If you do it without precision (by lumping things into a collection of related concepts, might be the simplest level) then composition is straightforward, and using multiple objects for the same relation is fine.

But if you do it with precision, you are building a much more complex model, because it incorporates all of the complexity of the existing models (by including them) and adds more layers of relations. That complexity is what I was referring to when I said it would be detailed—lots of details to make up a more complex model.
